x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1167,4 +1173,61 @@ public void accept(DiscoveryNode node, ClusterState state) {
     }
+
+    public static Path resolveConfigFile(Environment env, String name) {
+        final Path config = env.configFile().resolve(name);
+        final Path legacyConfig = env.configFile().resolve("x-pack").resolve(name);
+        // config and legacy config can be the same path if name is an absolute path
+        if (config.equals(legacyConfig) == false) {
+            final boolean configFileExists = Files.exists(config);
+            final boolean legacyConfigExists = Files.exists(legacyConfig);
+            if (configFileExists == false) {
+                if (legacyConfigExists) {
+                    deprecationLogger.deprecated("Config file [" + name + "] is in a deprecated location. Move from " +
+                        legacyConfig.toString() + " to " + config.toString());
+                    return legacyConfig;
+                }
+            } else if (legacyConfigExists) {
+                // there is a file in both locations
+                if (isDefaultFile(name, config)) {
+                    // use the legacy file as the new file is the default but warn user
+                    deprecationLogger.deprecated("Config file [" + name + "] exists in a deprecated location and non-deprecated location." +
+                        " The file in the non-deprecated location is the default file. Using file found in the deprecated location. Move " +
+                        legacyConfig.toString() + " to " + config.toString());
+                    return legacyConfig;
+                } else {
+                    // the regular file has been modified, but the old still exists, warn the user
+                    deprecationLogger.deprecated("Config file [" + name + "] exists in a deprecated location and non-deprecated location." +
+                        " Using file found in the non-deprecated location [" + config.toString() + "]. Determine which file should be" +
+                        " kept and move it to " + config.toString() + ", then remove " + legacyConfig.toString());
+                }
+            }
+        }
+        return config;
+    }
+
+    static boolean isDefaultFile(String name, Path file) {
+        InputStream in = null;
+        try {
+            in = XPackPlugin.class.getResourceAsStream("/config/" + name);
+            if (in != null) {
+                try (InputStream fin = Files.newInputStream(file)) {
+                    int inValue = in.read();
+                    int finValue = fin.read();
+                    while (inValue != -1 && finValue != -1 && inValue == finValue) {
+                        inValue = in.read();
+                        finValue = fin.read();
+                    }
+                    return inValue == finValue;
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } finally {
+            if (in != null) {
+                IOUtils.closeWhileHandlingException(in);
```

Comment:
What's the reason try-with-resources can't be used?

Reply:
I forgot that try-with-resources can handle nulls. Switched.